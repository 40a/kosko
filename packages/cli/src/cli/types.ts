import { Logger } from "./logger";

export interface WithAlias {
  alias?: string | string[];
}

export interface BaseOption<T = any> {
  description?: string;
  default?: T | (() => T | undefined);
  required?: boolean;
  options?: T[];
}

export function getAliases({ alias }: WithAlias): string[] {
  if (Array.isArray(alias)) return alias;
  return alias ? [alias] : [];
}

export interface Arg<K = string, V = any> extends BaseOption<V> {
  name: K;
}

export enum OptionType {
  String = "string",
  Boolean = "boolean",
  Count = "count",
  Number = "number",
  Array = "array"
}

export interface Option<T = any> extends WithAlias, BaseOption<T> {
  type?: OptionType;
  group?: string;
}

export interface CommandMeta<Options = any> {
  usage?: string;
  description?: string;
  examples?: string;
  options?: { [key in keyof Options]: Option };
  commands?: { [key: string]: Command };
  args?: Arg[];
}

export interface Command<Options = any> extends CommandMeta<Options> {
  exec(ctx: Context, argv: string[]): void;
}

export interface Context {
  logger: Logger;
}
