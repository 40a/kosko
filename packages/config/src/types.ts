export interface EnvironmentConfig {
  /** External modules to require. */
  require?: ReadonlyArray<string>;
  /** Component patterns. */
  components?: ReadonlyArray<string>;
}

export interface Config extends EnvironmentConfig {
  /** Environment configs. */
  environments?: { [key: string]: EnvironmentConfig };
}
