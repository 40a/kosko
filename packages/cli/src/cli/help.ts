import groupBy from "lodash.groupby";
import { Writable } from "stream";
import { Option, OptionType, CommandMeta } from "./types";

interface Entry {
  key: string;
  description: string;
}

interface OptionWithName extends Option {
  name: string;
}

function getOptionKey(opt: OptionWithName) {
  let key = [opt.name]
    .concat(opt.alias || [])
    .map(x => (x.length > 1 ? `--${x}` : `-${x}`))
    .sort((a, b) => a.length - b.length)
    .join(", ");

  switch (opt.type) {
    case OptionType.Boolean:
    case OptionType.Count:
      break;

    default:
      if (opt.type) key += ` ${opt.type}`;
  }

  return key;
}

class Help {
  private readonly cmd: CommandMeta;
  private readonly writer: Writable;

  private started = false;

  constructor(cmd: CommandMeta, writer: Writable) {
    this.cmd = cmd;
    this.writer = writer;
  }

  public async print() {
    await this.printUsage();
    await this.printDescription();
    await this.printArgs();
    await this.printCommands();
    await this.printOptions();
  }

  private write(chunk: any) {
    return new Promise((resolve, reject) => {
      this.writer.write(chunk, err => {
        if (err) return reject(err);
        resolve();
      });
    });
  }

  private async printUsage() {
    const { usage } = this.cmd;
    if (!usage) return;

    await this.printHeader("Usage: " + usage);
  }

  private async printDescription() {
    const { description } = this.cmd;
    if (!description) return;

    await this.printHeader(description);
  }

  private async printArgs() {
    const { args = [] } = this.cmd;

    await this.printEntries(
      "Arguments:",
      args.map(arg => ({ key: arg.name, description: arg.description || "" }))
    );
  }

  private async printCommands() {
    const { commands = {} } = this.cmd;
    const entries: Entry[] = [];

    for (const key of Object.keys(commands)) {
      entries.push({ key, description: commands[key].description || "" });
    }

    await this.printEntries("Commands:", entries);
  }

  private async printOptions() {
    const { options = {} } = this.cmd;
    const opts: OptionWithName[] = [];

    for (const key of Object.keys(options)) {
      opts.push({ ...options[key], name: key });
    }

    const groups = groupBy(opts, opt => opt.group || "Options");

    for (const key of Object.keys(groups)) {
      await this.printOptionGroup(key, groups[key]);
    }
  }

  private async printOptionGroup(group: string, options: OptionWithName[]) {
    await this.printEntries(
      `${group}:`,
      options.map(opt => ({
        key: getOptionKey(opt),
        description: opt.description || ""
      }))
    );
  }

  private async printHeader(header?: string) {
    if (this.started) await this.write("\n");
    await this.write(header + "\n");
    this.started = true;
  }

  private async printEntries(header: string, entries: Entry[]) {
    if (!entries.length) return;

    await this.printHeader(header);

    const keySize = Math.max(...entries.map(e => e.key.length));

    entries.sort((a, b) => {
      if (a.key > b.key) return 1;
      if (a.key < b.key) return -1;
      return 0;
    });

    for (const entry of entries) {
      const pad = Array(keySize - entry.key.length)
        .fill(" ")
        .join("");

      await this.write(`  ${entry.key}${pad}  ${entry.description}\n`);
    }
  }
}

export async function help<T = any>(
  cmd: CommandMeta<T>,
  writer: Writable = process.stdout
) {
  const helper = new Help(cmd, writer);
  return helper.print();
}
