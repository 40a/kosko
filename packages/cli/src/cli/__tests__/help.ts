import { help } from "../help";
import { CommandMeta, OptionType } from "../types";
import BufferList from "bl";

let cmd: CommandMeta;
let bl: BufferList;

beforeEach(async () => {
  bl = new BufferList();
  await help(cmd, bl);
});

describe("usage", () => {
  beforeAll(() => {
    cmd = { usage: "foo bar" };
  });

  test("should print the usage", () => {
    expect(bl.toString()).toContain(`Usage: ${cmd.usage}`);
  });
});

describe("description", () => {
  beforeAll(() => {
    cmd = { description: "do something" };
  });

  test("should print the description", () => {
    expect(bl.toString()).toContain(cmd.description);
  });
});

describe("args", () => {
  beforeAll(() => {
    cmd = {
      args: [
        { name: "foo", description: "desc of foo" },
        { name: "bazz", description: "desc of bazz" }
      ]
    };
  });

  test("should print the arguments", () => {
    expect(bl.toString()).toMatchSnapshot();
  });
});

describe("commands", () => {
  beforeAll(() => {
    cmd = {
      commands: {
        foo: { description: "desc of foo" },
        bazz: { description: "desc of bazz" }
      } as any
    };
  });

  test("should print the commands", () => {
    expect(bl.toString()).toMatchSnapshot();
  });
});

describe("options", () => {
  beforeAll(() => {
    cmd = {
      options: {
        str: { type: OptionType.String, description: "desc of str" },
        bool: { type: OptionType.Boolean, description: "desc of bool" },
        count: { type: OptionType.Count, description: "desc of count" },
        arr: { type: OptionType.Array, description: "desc of arr" },
        num: { type: OptionType.Number, description: "desc of num" },
        unknown: { description: "some unknown option" },
        onealias: { alias: "o", description: "one alias" },
        morealias: { alias: ["m", "more"], description: "more aliases" },
        camelCase: { description: "camel case" }
      }
    };
  });

  test("should print the options", () => {
    expect(bl.toString()).toMatchSnapshot();
  });
});

describe("options with groups", () => {
  beforeAll(() => {
    cmd = {
      options: {
        help: { description: "show help", group: "Global Options" },
        foo: { description: "foo" },
        version: { description: "show version", group: "Global Options" },
        bar: { description: "bar", group: "Hidden Options" }
      }
    };
  });

  test("should print the options", () => {
    expect(bl.toString()).toMatchSnapshot();
  });
});

describe("put all together", () => {
  beforeAll(() => {
    cmd = {
      usage: "foo",
      description: "desc foo desc",
      commands: {
        bar: { description: "subcommand bar" } as any
      },
      options: {
        foo: { description: "option foo" }
      },
      args: [{ name: "baz", description: "arg bar" }]
    };
  });

  test("should print the options", () => {
    expect(bl.toString()).toMatchSnapshot();
  });
});
