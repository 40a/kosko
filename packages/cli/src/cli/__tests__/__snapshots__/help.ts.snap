// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`args should print the arguments 1`] = `
"Arguments:
  bazz  desc of bazz
  foo   desc of foo
"
`;

exports[`commands should print the commands 1`] = `
"Commands:
  bazz  desc of bazz
  foo   desc of foo
"
`;

exports[`options should print the options 1`] = `
"Options:
  --arr array              desc of arr
  --bool                   desc of bool
  --count                  desc of count
  --num number             desc of num
  --str string             desc of str
  --unknown                some unknown option
  -m, --more, --morealias  more aliases
  -o, --onealias           one alias
"
`;

exports[`options with groups should print the options 1`] = `
"Global Options:
  --help     show help
  --version  show version

Options:
  --foo  foo

Hidden Options:
  --bar  bar
"
`;

exports[`put all together should print the options 1`] = `
"Usage: foo

desc foo desc

Arguments:
  baz  arg bar

Commands:
  bar  subcommand bar

Options:
  --foo  option foo
"
`;
